contracts/proxy_contract/v1/ergoscript/phoenix_v1_hodlcoin_proxy.es

```diff
@@ -185,39 +184,48 @@
         }
 
     } else {
-        
-        // ===== Refund Tx ===== //
-        val validRefundTx Boolean = {
 
-            // Outputs
-            val buyerPKBoxOUT: Box = OUTPUTS(0)
-            val minerFeeBoxOUT: Box = OUTPUTS(1)
+        // ===== Refund Tx ===== //
+        val validRefundTx: Boolean = {
+
+            val recommendedMinerFee: Long = 1000000L // overrides buyer's miner fee choice since they could go too low or too high
+
+            //ensures buyer receives total value of box
+            val validValueTransfer: Boolean = OUTPUTS.map { (o: Box) =>
+                if (o.propositionBytes == buyerPK.propBytes) o.value else 0L
+            }.fold(0L, { (a: Long, b: Long) => a + b }) >= SELF.value
+
+            // if box has tokens it must go to buyer
+            val validTokenTransfer: Boolean = {
+                if(SELF.tokens.size > 0){
+                    OUTPUTS.exists { (o: Box) =>
+                        (o.tokens == SELF.tokens) && (o.propositionBytes == buyerPK.propBytes)
+                    }
+                } else{
+                  true
```

Comment:
this is tautological, a refund will always have a token in it

Reply:
loop unnecessary

Reply:
this is needed if there are multiple input boxes with various tokens